design/deploy-steps.md

```diff
@@ -0,0 +1,142 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# Customizable deployment procedure
+
+## Status
+
+implementable
+
+## Summary
+
+This design proposes customizing the deployment procedure by providing a high
+level abstraction around Ironic [deploy steps][deploy-steps], as well as
+allowing per-host deployment (IPA) images.
+
+[deploy-steps]: https://docs.openstack.org/ironic/latest/admin/node-deployment.html#deploy-steps
+
+## Motivation
+
+By default the Ironic agent downloads a provided image via HTTP(s), converts it
+(if needed) and writes directly to the disk. This procedure works for many
+distributions, but may not suit some cases well. A prominent example is CoreOS,
+which has its own installer ([coreos-installer][coreos-installer]).
+
+[coreos-installer]: https://github.com/coreos/coreos-installer
+
+### Goals
+
+- A way to replace the default deploy procedure with a custom action.
+
+### Non-Goals
+
+- Expose the full functionality of Ironic deploy steps.
+- Add a way to execute custom actions during deployment.
+- A way to provide a link to the Ironic agent image per host.
+
+## Proposal
+
+### User Story
+
+As a CoreOS user I want to use `coreos-installer` to write the CoreOS image
+without losing any functionality that Ironic offers (unlike the live ISO
+workflow).
+
+## Design Details
+
+Additions to `BareMetalHost`:
+
+- New structure field `CustomDeploy` with the only field `Method` to engage
+  a non-default deployment procedure as described below.
+- The `Image.URL` field will become optional if `CustomDeploy` is set
+  since we don't know if the deploy method actually requires a separate image.
+
+Setting `CustomDeploy` to a non-empty value will trigger deployment even if
+`Image.URL` is empty.
+
+### Implementation Details: CustomDeploy
+
+When `CustomDeploy` is set, we will use the new [custom-agent deploy
+interface][custom-agent] so that the default deploy steps do not apply. Then we
+request a deploy step with the name provided in the `CustomDeploy.Method`
+field. This step has to be provided in the agent image and is responsible for
+the whole deployment process.
+
+NOTE: `CustomDeploy` is a single-field structure to allow potential extensions
```

Comment:
Yeah, what Zane said. Having one very generic `Method` may also allow us to accommodate things like `ansible` deploy.

Reply:
Is there a reason to not make it a `[]string` -- sort of in the middle between the two approaches? I understand supporting arguments complicates things, but limiting ourselves to just one method makes it harder to change the API later.

Reply:
It is normal, yes. But without arguments we cannot reuse upstream steps anyway. If we write our own steps, it does not matter if there is one or many.

Having a list of steps exposes substantially more about the underlying implementation, I guess that's Zane's concern.

Reply:
I'd be more inclined to keep the feature limited to one (deploy) step only, to avoid any over-complication and limiting its misuse, since the pre-requisites are already non-trivial for just a single step (the user must known in advance the correct step name, how to pass/format any required data via network/meta/userData, and I suppose the step itself must be available somehow within Ironic)

Reply:
Ok you've convinced me